modulemd/v2/modulemd-module-stream.c

```diff
@@ -0,0 +1,346 @@
+/*
+ * This file is part of libmodulemd
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * Fedora-License-Identifier: MIT
+ * SPDX-2.0-License-Identifier: MIT
+ * SPDX-3.0-License-Identifier: MIT
+ *
+ * This program is free software.
+ * For more information on the license, see COPYING.
+ * For more information on free software, see <https://www.gnu.org/philosophy/free-sw.en.html>.
+ */
+
+#include "modulemd-module-stream.h"
+#include "modulemd-module-stream-v1.h"
+#include "modulemd-module-stream-v2.h"
+#include "private/modulemd-module-stream-private.h"
+#include "private/modulemd-util.h"
+
+typedef struct
+{
+  guint64 mdversion;
+  gchar *module_name;
+  gchar *stream_name;
+} ModulemdModuleStreamPrivate;
+
+G_DEFINE_ABSTRACT_TYPE_WITH_PRIVATE (ModulemdModuleStream,
+                                     modulemd_module_stream,
+                                     G_TYPE_OBJECT)
+
+enum
+{
+  PROP_0,
+  PROP_MDVERSION,
+  PROP_MODULE_NAME,
+  PROP_STREAM_NAME,
+  N_PROPS
+};
+
+static GParamSpec *properties[N_PROPS];
+
+ModulemdModuleStream *
+modulemd_module_stream_new (guint64 mdversion,
+                            const gchar *module_name,
+                            const gchar *stream_name)
+{
+  switch (mdversion)
+    {
+    case MD_MODULESTREAM_VERSION_ONE:
+      return MODULEMD_MODULE_STREAM (
+        modulemd_module_stream_v1_new (module_name, stream_name));
+
+    case MD_MODULESTREAM_VERSION_TWO:
+      return MODULEMD_MODULE_STREAM (
+        modulemd_module_stream_v2_new (module_name, stream_name));
+
+    default:
+      /* Other versions have not yet been implemented */
+      return NULL;
+    }
+}
+
+
+static void
+modulemd_module_stream_finalize (GObject *object)
+{
+  ModulemdModuleStream *self = (ModulemdModuleStream *)object;
+  ModulemdModuleStreamPrivate *priv =
+    modulemd_module_stream_get_instance_private (self);
+
+  g_clear_pointer (&priv->module_name, g_free);
+  g_clear_pointer (&priv->stream_name, g_free);
+
+  G_OBJECT_CLASS (modulemd_module_stream_parent_class)->finalize (object);
+}
+
+
+static ModulemdModuleStream *
+modulemd_module_stream_default_copy (ModulemdModuleStream *self,
+                                     const gchar *stream_name)
+{
+  const gchar *stream = NULL;
+
+  /* If the stream name was passed in, replace the original with it */
+  if (stream_name)
+    {
+      stream = stream_name;
+    }
+  else
+    {
+      stream = modulemd_module_stream_get_stream_name (self);
+    }
+
+  /* The base implementation of the copy */
+  return modulemd_module_stream_new (
+    modulemd_module_stream_get_mdversion (self),
+    modulemd_module_stream_get_module_name (self),
+    stream);
+}
+
+
+ModulemdModuleStream *
+modulemd_module_stream_copy (ModulemdModuleStream *self,
+                             const gchar *stream_name)
+{
+  ModulemdModuleStreamClass *klass;
+
+  g_return_val_if_fail (MODULEMD_IS_MODULE_STREAM (self), NULL);
+
+  if (!self)
```

Comment:
Shouldn't this be caught by the `g_return_val_if_fail()` call?

Reply:
Incorrect ordering. The NULL check should return before `g_return_val_if_fail()` because that's a valid request (copy nothing, get nothing).